ledger/src/advance.rs

```diff
@@ -127,30 +129,46 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
                 // Retrieve the latest epoch challenge.
                 let latest_epoch_challenge = self.latest_epoch_challenge()?;
-                // TODO: For mainnet - Add `aborted_solution_ids` to the block. And optimize this logic.
-                // Verify the candidate solutions.
-                let verification_results: Vec<_> = cfg_into_iter!(candidate_solutions)
-                    .map(|solution| {
-                        (
-                            solution,
-                            solution
-                                .verify(coinbase_verifying_key, &latest_epoch_challenge, self.latest_proof_target())
-                                .unwrap_or(false),
-                        )
-                    })
-                    .collect();
-                // Separate the candidate solutions into valid and aborted solutions.
-                let mut valid_candidate_solutions = Vec::with_capacity(N::MAX_SOLUTIONS);
-                let mut aborted_candidate_solutions = Vec::new();
-                for (solution, is_valid) in verification_results.into_iter() {
-                    if is_valid && valid_candidate_solutions.len() < N::MAX_SOLUTIONS {
-                        valid_candidate_solutions.push(solution);
-                    } else {
-                        aborted_candidate_solutions.push(solution);
+                // TODO: For mainnet - Add `aborted_solution_ids` to the block.
+                let valid_candidate_solutions = RwLock::new(Vec::with_capacity(N::MAX_SOLUTIONS));
+                let aborted_candidate_solutions = Mutex::new(Vec::new());
+
+                // Separate the candidate solutions into valid and aborted ones.
+                cfg_into_iter!(candidate_solutions).for_each(|solution| {
+                    // Check if there are N::MAX_SOLUTIONS valid solutions already; this
+                    // is an equality check, because that number can't be breached.
+                    let limit_reached = valid_candidate_solutions.read().len() == N::MAX_SOLUTIONS;
+                    if limit_reached {
+                        // If that's the case, mark the solution as aborted.
+                        aborted_candidate_solutions.lock().push(solution);
+                        return;
                     }
-                }
+
+                    // If there is still room for solutions, verify the next candidate.
+                    let verified = solution
+                        .verify(coinbase_verifying_key, &latest_epoch_challenge, self.latest_proof_target())
+                        .unwrap_or(false);
+
+                    match verified {
+                        true => {
+                            // Since the solutions are processed in parallel, make sure that the
+                            // collection of the valid ones is still below limit before pushing.
+                            let mut valid_candidate_solutions = valid_candidate_solutions.write();
```

Comment:
note: I used `write` instead of a `read` and a conditional `write`, as we're expecting to be below the limit due to the earlier ([L141](https://github.com/AleoHQ/snarkVM/pull/2256/files#diff-bfbe2e2f8e1e6a889781c83d0ae44ba3dde3604d9415aaf27921d7c94acd02c6R141)) length check